addons/persistence/postgresql-persistence-addon/src/main/java/org/kie/kogito/persistence/postgresql/PostgreProcessInstances.java

```diff
@@ -74,48 +83,92 @@ public boolean exists(String id) {
         return findById(id).isPresent();
     }
 
+    @SuppressWarnings("unchecked")
     @Override
     public void create(String id, ProcessInstance instance) {
+        if (!isActive(instance)) {
+            disconnect(instance);
+            return;
+        }
         insertInternal(UUID.fromString(id), marshaller.marshallProcessInstance(instance));
         disconnect(instance);
     }
 
+    @SuppressWarnings("unchecked")
     @Override
     public void update(String id, ProcessInstance instance) {
-        updateInternal(UUID.fromString(id), marshaller.marshallProcessInstance(instance));
+        if (!isActive(instance)) {
+            disconnect(instance);
+            return;
+        }
+        if (lock) {
+            updateWithLock(UUID.fromString(id), marshaller.marshallProcessInstance(instance), instance.version());
+        } else {
+            updateInternal(UUID.fromString(id), marshaller.marshallProcessInstance(instance));
+        }
         disconnect(instance);
     }
 
     @Override
     public void remove(String id) {
-        deleteInternal(UUID.fromString(id));
+        boolean isDeleted = deleteInternal(UUID.fromString(id));
+        if (lock && !isDeleted) {
+            throw uncheckedException(null, "The document with ID: %s was updated or deleted by other request.", id);
+        }
     }
 
     @Override
     public Optional<ProcessInstance> findById(String id, ProcessInstanceReadMode mode) {
-        return findByIdInternal(UUID.fromString(id)).map(b -> marshaller.unmarshallProcessInstance(b, process));
+        Optional<Row> row = findByIdInternal(UUID.fromString(id));
+
+        if (row.isPresent()) {
+            Optional<byte[]> payload = row
+                    .map(r -> r.getBuffer(PAYLOAD))
+                    .map(Buffer::getBytes);
+            if (payload.isPresent()) {
+                ProcessInstance<?> instance = mode == MUTABLE ? marshaller.unmarshallProcessInstance(payload.get(),
+                        process)
+                        : marshaller.unmarshallReadOnlyProcessInstance(payload.get(), process);
+                ((AbstractProcessInstance) instance).setVersion(row.get().getLong(VERSION));
+                return Optional.ofNullable(instance);
```

Comment:
```suggestion
                return Optional.of(instance);
```
as it cannot be null, otherwise the previous `setVersion` would fail.

Reply:
done.